graphql/batch_executor.go

```diff
@@ -0,0 +1,409 @@
+	// result values.  Filter out invalid/nil objects.
+	nonNilSources := make([]interface{}, 0, len(sources))
+	nonNilDestinations := make([]map[string]interface{}, 0, len(destinations))
+	originDestinations := make([]*objectWriter, 0, len(destinations))
+	for idx, source := range sources {
+		value := reflect.ValueOf(source)
+		if value.Kind() == reflect.Ptr && value.IsNil() {
+			destinations[idx].Fill(nil)
+			continue
+		}
+		nonNilSources = append(nonNilSources, source)
+		destMap := make(map[string]interface{}, len(selections))
+		destinations[idx].Fill(destMap)
+		nonNilDestinations = append(nonNilDestinations, destMap)
+		originDestinations = append(originDestinations, destinations[idx])
+	}
+
+	// Number of Execution Units = (NumExpensiveFields x NumSources) + NumNonExpensiveFields
+	executionUnits := make([]*ExecutionUnit, 0, numNonExpensive+(numExpensive*len(nonNilSources)))
+
+	// for every selection, resolve the value or schedule an execution unit for the field
+	for _, selection := range selections {
+		if selection.Name == "__typename" {
+			for idx := range nonNilDestinations {
+				nonNilDestinations[idx][selection.Alias] = typ.Name
+			}
+			continue
+		}
+
+		destForSelection := make([]*objectWriter, 0, len(nonNilDestinations))
+		for idx, destMap := range nonNilDestinations {
+			filler := newObjectWriter(originDestinations[idx], selection.Alias)
+			destForSelection = append(destForSelection, filler)
+			destMap[selection.Alias] = filler
+		}
+
+		field := typ.Fields[selection.Name]
+		if field.Expensive {
+			// Expensive fields should be executed as multiple "Units".  The scheduler
+			// controls how the units are executed
+			for idx, source := range nonNilSources {
+				executionUnits = append(executionUnits, &ExecutionUnit{
+					ctx:          ctx,
+					field:        field,
+					sources:      []interface{}{source},
+					destinations: []*objectWriter{destForSelection[idx]},
+					selection:    selection,
+				})
+			}
+			continue
+		}
+		if field.External {
+			// External non-Expensive fields should be fast (so we can run them at the
+			// same time), but, since they are still external functions we don't want
+			// to run them where they could potentially block.
+			// So we create an execution unit with all the fields to execute
+			// asynchronously.
+			executionUnits = append(executionUnits, &ExecutionUnit{
+				ctx:          ctx,
+				field:        field,
+				sources:      nonNilSources,
+				destinations: destForSelection,
+				selection:    selection,
+			})
+			continue
+		}
+		// If the fields are not expensive or external the execution time should be
+		// bounded, so we can resolve them immediately.
+		executionUnits = append(
+			executionUnits,
+			resolveExecutionUnit(&ExecutionUnit{
+				ctx:          ctx,
+				field:        field,
+				sources:      nonNilSources,
+				destinations: destForSelection,
+				selection:    selection,
+			})...,
+		)
+	}
+
+	if typ.KeyField != nil {
```

Comment:
could we get this in the selection somehow? would be nice to have the same expensive etc logic for these

Reply:
We could probably do it at the parsing layer and add it to the selection set based on the same parameters we're using here.

Reply:
Maybe a follow-up?